Bell API

Lama06 opened the pull request:
Adds methods to ring a bell and to query its state

Lama06 committed:
Bell API

Lama06 committed:
Return null from getRingingDirection if the bell is not ringing

Lama06 committed:
Rebase

Lulu13022002 commented on patches/server/0959-Bell-API.patch:
```diff
@@ -0,0 +1,62 @@
++    public org.bukkit.block.BlockFace getRingingDirection() {
++        if (!isRinging()) return null;
++
++        return switch (getTileEntity().clickDirection) {
```
You can use the CraftBlock.notchToBlockFace method to avoid that. Also it's a little bit strange that you have mapped the up/down direction since you explicitly blocked that in the ring method

Lulu13022002 commented on patches/server/0959-Bell-API.patch:
```diff
@@ -0,0 +1,62 @@
++        if (!direction.isCartesian()) throw new IllegalArgumentException("A bell can only ring in a cartesian direction");
++        if (direction.getModY() != 0) throw new IllegalArgumentException("A bell cannot ring up or down");
++        requirePlaced();
++        getTileEntity().onHit(net.minecraft.core.Direction.fromNormal(direction.getModX(), direction.getModY(), direction.getModZ()));
```
Since you have checked that the block face is cartesian, you can directly use the CraftBlock.blockFaceToNotch

Lama06 replied:
Thanks for your feedback, I now use these methods

Lama06 committed:
Use CraftBlock.blockFaceToNotch and CraftBlock.notchToBlockFace

Lama06 committed:
Rebase

Machine-Maker commented on patches/api/0429-Bell-API.patch:
```diff
@@ -0,0 +1,67 @@
++
++    /**
++     * Returns if the bell is currently resonating.
++     * The resonation sound starts playing 5 ticks after the bell is hit and ends 40 ticks later.
```
This also requires nearby raiders.

Machine-Maker commented on patches/api/0429-Bell-API.patch:
```diff
@@ -0,0 +1,67 @@
++     * A bell stops ringing after 50 ticks.
++     * @return The number of ticks this bell has been ringing for
++     */
++    int getHasBeenRingingFor();
```
```diff
-    int getHasBeenRingingFor();
+    int getTicksRinging();
```
or something, should have some mention of a time or ticks in the method name.

Machine-Maker commented on patches/api/0429-Bell-API.patch:
```diff
@@ -0,0 +1,67 @@
++     * A bell stops resonating after 40 ticks.
++     * @return The number of ticks this bell has been resonating for
++     */
++    int getHasBeenResonatingFor();
```
same concern as comment on `getHasBeenRingingFor`

Machine-Maker commented on patches/api/0429-Bell-API.patch:
```diff
@@ -0,0 +1,67 @@
++     */
++    default void ring() {
++        ring(((org.bukkit.block.data.type.Bell) getBlockData()).getFacing());
++    }
```
I don't think this should be a default method, it just is an unneeded conversion to API types when you can just do it in the impl without any API types created.

Lama06 replied:
Thanks for all your feedback! I have done everything you suggested now.

Lama06 committed:
Rebase
Make Bell#ring() an abstract method
Rename Bell#getHasBeenRingingFor() to Bell#getTicksRinging()
Rename Bell#getHasBennResonatingFor() to Bell#getTicksResonating()
Mention that a bell only resonates if there are raiders nearby in the javadocs

lynxplay commented on patches/server/0962-Bell-API.patch:
```diff
@@ -0,0 +1,61 @@
++
++    @Override
++    public int getTicksResonating() {
++        return getTileEntity().resonationTicks;
```
I have a bit of an issue with this method.
`resonationTicks` is not reset to 0 until the next time the bell is rang.
A non resonating bell will end up with a 40 tick val on this, even if it is no longer resonating.

Lama06 replied:
I now check whether the bell is still resonating and return 0 if it is not

lynxplay commented on patches/api/0418-Bell-API.patch:
```diff
@@ -0,0 +1,65 @@
++     * Returns the direction the bell is ringing in or null if it is not ringing.
++     * @return The direction the bell is ringing in
++     */
++    @org.jetbrains.annotations.Nullable BlockFace getRingingDirection();
```
This method needs a bit more javadocs.
The ringing direction is not really the direction a player rang the bell in, it is the relative block face the bell was hit on.

If I stand infront of a bell that is `facing=west` I can ring it in, as a player, in the directions EAST and WEST.
When I look in the EAST direction and I "ring" the bell, this method returns "WEST" as I hit it while looking east.

If this is a linguistic thing or not (a west wind for example is a wind that blows east etc, so a bell rining from the west to the east could be interpreted as such) does not matter I think, more documentation is required here to ensure we don't run into confusion about the returned value.

Lama06 replied:
I have now added javadocs to this method and Bell#ring(BlockFace) to make clear what is meant by the direction a bell is ringing in.
If my javadocs sound weird please correct me, I am not a native English speaker

lynxplay commented on patches/api/0418-Bell-API.patch:
```diff
@@ -0,0 +1,65 @@
++    /**
++     * Rings the bell in the direction that it is facing.
++     */
++    void ring();
```
This method also requires more documentation imo detailing something that, at least `ring` does imply.
The lack of sound when calling this method.

The same goes for the below method, neither will cause a sound to be played.
While this is "understandable" for developers deep in the server code, it should be documented to avoid confusion.

Lama06 replied:
I now mention this in the javadocs of this method.
Thanks for all your feedback!

Lama06 committed:
Rebase

Lama06 committed:
Improve Javadocs

Lama06 committed:
Make Bell#getTicksResonating() return 0 if the bell has stopped resonating
